src/v2/Artsy/Relay/middleware/cache/Cache.ts

```diff
@@ -0,0 +1,166 @@
+import createLogger from "v2/Utils/logger"
+import RelayQueryResponseCache from "relay-runtime/lib/network/RelayQueryResponseCache"
+import { isServer } from "lib/environment"
+import { RetryStrategyOptions } from "redis"
+
+import {
+  GraphQLResponse,
+  QueryResponseCache,
+  CacheConfig as RelayCacheConfig,
+  Variables as RelayVariables,
+} from "relay-runtime"
+
+const logger = createLogger("v2/Artsy/middleware/cache/Cache")
+
+export interface CacheConfig {
+  size: number
+  ttl: number
+  disableServerSideCache?: boolean
+}
+
+export class Cache {
+  cacheConfig: CacheConfig
+  enableServerSideCache: boolean
+  relayCache: RelayQueryResponseCache
+
+  redisCache: {
+    del: (key: string) => Promise<void>
+    expire: (key: string, ttl: number) => Promise<void>
+    flushall: () => Promise<void>
+    get: (key: string) => Promise<string>
+    set: (
+      key: string,
+      value: string,
+      expirationCommandCode?: string,
+      expirationValue?: string | number
+    ) => Promise<void>
+  }
+
+  constructor(cacheConfig: CacheConfig) {
+    this.cacheConfig = cacheConfig
+    this.enableServerSideCache =
+      isServer && !this.cacheConfig.disableServerSideCache
+    this.initRelayCache()
+    this.initRedisCache()
+  }
+
+  initRelayCache() {
+    this.relayCache = new QueryResponseCache(this.cacheConfig)
+  }
+
+  initRedisCache() {
+    if (!this.enableServerSideCache) {
+      return
+    }
+
+    const { promisify } = require("util")
+    const redis = require("redis")
+
+    const client = redis.createClient({
+      url: process.env.OPENREDIS_URL,
+      retry_strategy: this.handleRedisError,
+    })
+
+    this.redisCache = {
+      del: promisify(client.del).bind(client),
+      expire: promisify(client.expire).bind(client),
+      flushall: promisify(client.flushall).bind(client),
+      get: promisify(client.get).bind(client),
+      set: promisify(client.set).bind(client),
+    }
+  }
+
+  handleRedisError(retryOptions: RetryStrategyOptions) {
```

Comment:
Would you move this to its own function. It doesn't really belong on the `Cache` class as it's not usable outside of the redis client itself.